File: Java/Leetcode695.java
package Java;

import java.util.LinkedList;
import java.util.Queue;

public class Leetcode695 {

    int count = 0;

    public int maxAreaOfIsland(int[][] grid){
        if(grid == null || grid.length == 0 || grid[0].length == 0) {return 0;}

        int row = grid.length, column = grid[0].length;
        int maxLand = 0;

        for(int r=0; r<row ; r++){
            for(int c=0; c<column; c++){
                if(grid[r][c] == 1){
                    count = 0;
                    dfs(grid, r, c);
                    // bfs(grid, r, c);
                    maxLand = Math.max(maxLand, count);
                }
            }
        }
        return maxLand;
    }

    public void dfs(int[][] grid, int r, int c){
        if(r<0 || c<0 || r>=grid.length || c>=grid[0].length || grid[r][c] == 0) {return;}
        count += 1;
        grid[r][c] = 0;
        dfs(grid, r-1, c);
        dfs(grid, r, c-1);
        dfs(grid, r+1, c);
        dfs(grid, r, c+1);
    }

    public void bfs(int[][] grid, int r, int c){
        int[] dr = {0, -1, 0, 1};
        int[] dc = {-1, 0, 1, 0};

        Queue<int[]> queue = new LinkedList<>();

        grid[r][c] = 0;
        count += 1;

        queue.add(new int[]{r, c});

        while(!queue.isEmpty()){
            // poll: get value from queue
            int[] temp = queue.poll();
            int tempR = temp[0];
            int tempC = temp[1];

            for(int i=0; i<4; i++){
                int curR = tempR + dr[i];
                int curC = tempC + dc[i];

                if(curR >=0 && curC >=0 && curR < grid.length && curC < grid[0].length && grid[curR][curC] == 1){
                    count += 1;
                    grid[curR][curC] = 0;
                    queue.add(new int[]{curR, curC});
                }
            }

        }

    }
}

File: Python/Leetcode695.py
class Leetcode695:
    def maxAreaOfIsland(self, grid: List[List[int]]) -> int:
        m, n = len(grid), len(grid[0])
        dx = [-1, 0, 1, 0]
        dy = [0, -1, 0, 1]

        def dfs(r, c):
            # 1이 아니면 return 0
            if r<0 or r==m or c<0 or c==n or grid[r][c]==0:
                return 0
            # dfs[r][c]가 1이면 우선 answer 1로 설정 후 해당 영역 0으로 바꿈
            answer = 1
            grid[r][c] = 0
            # 재귀적으로 주변 땅 탐색, 주변 땅에 1이 있으면 기존 answer에 더해지고
            # 만약 맨 위의 if문에 걸리면 0이 되어서 더해지지 않음
            for i in range(4):
                answer += dfs(r+dx[i], c+dy[i])
            return answer

        answer = 0
        for r in range(m):
            for c in range(n):
                answer = max(answer, dfs(r,c))
        return answer

        def bfs(r, c):
            # queue에 삽입 후 0으로 바꿈
            q = deque([(r,c)])
            grid[r][c] = 0

            answer = 0
            while q:
                r, c = q.popleft()
                answer += 1
                for i in range(4):
                    nr, nc = r+dx[i], c+dy[i]
                    if nr<0 or nr==m or nc<0 or nc==n or grid[nr][nc]==0: continue
                    grid[nr][nc] = 0
                    q.append(nr, nc)
            return answer

        answer = 0
        for r in range(m):
            for c in range(n):
                if grid[r][c] == 1:
                    answer = max(answer, bfs(r, c))
        
        return answer